### PR TITLE
options: splitting into base class and command line parsing

### PR DESCRIPTION
**Pull request by alyssawilk**
This allows removing the tclap dep from Envoy Mobile.

Risk Level: low
Testing: n/a
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by alyssawilk** [comment]
This is simply a code split with the exception of the new absl:::Status setLogLevel(absl::string_view) wrapper to avoid a bunch of boilerplate in E-M

File: source/server/options_impl.cc [hunk 1]
```diff
@@ -214,7 +214,7 @@ OptionsImpl::OptionsImpl(std::vector<std::string> args,
     mode_ = Server::Mode::InitOnly;
   } else {
     const std::string message = fmt::format("error: unknown mode '{}'", mode.getValue());
-    throwMalformedArgExceptionOrPanic(message);
+    throw MalformedArgvException(message);
```
Comment (wbpcode) [inline source/server/options_impl.cc:217]: do these changes make sense or just a cleanup?

**Review by wbpcode — COMMENTED** [review]
Thanks for doing this and two comments are added.

If I got it correctly, this PR split previous `OptionsImpl` to `OptionsImplBase` to store all bootstrap options and new `OptionsImpl` to parse command line by the TCLAP. And and parsing part is unnecessary for the E-M, so we can use `OptionsImplBase` only in the E-M.

**Comment by wbpcode** [comment]
And please check the CI, thanks.

/wait

File: source/server/options_impl.cc [hunk 2]
```diff
@@ -194,7 +190,11 @@ OptionsImpl::OptionsImpl(std::vector<std::string> args,
   cpuset_threads_ = cpuset_threads.getValue();
 
   if (log_level.isSet()) {
-    log_level_ = parseAndValidateLogLevel(log_level.getValue());
+    auto status_or_error = parseAndValidateLogLevel(log_level.getValue());
+    if (!status_or_error.status().ok()) {
+      logError(std::string(status_or_error.status().message()));
+    }
+    log_level_ = status_or_error.value();
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/server/options_impl.cc:197]: logError throws an exception, so we won't hit this if value is not OK.

File: source/server/options_impl.cc (hunk 1 above)
Comment (alyssawilk) [reply source/server/options_impl.cc:217]: the OrPanic macros are needed for E-M exceptionless build.  now this code is not used by E-M we can go back to the original throw.

**Review by alyssawilk — COMMENTED** [review]
correct about the split.  CI looks weird - a bunch of jobs randomly canceled.  retesting but lmk if you want any changes!

**Comment by alyssawilk** [comment]
oops there were 2 build fails buried in the cancels.  on it!

**Comment by wbpcode** [comment]
oh, seems I cannot stamp it today before I go to 🛏️. CI still say no.

**Comment by alyssawilk** [comment]
Azure Pipelines / envoy-presubmit (Linux x64 Build and test)
Started 19h 19m 57s ago
well that's not great.

**Comment by alyssawilk** [comment]
ugh I have no idea what's with the coverage failure and can't repro locally.  I'll ping when this is ready for review.

**Comment by wbpcode** [comment]
🤣 Add waiting label first.

**Comment by alyssawilk** [comment]
FINALLY GREEN

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**